In [ ]:
# Multiple Linear Regression

# Data Inspection : Student Performance Dataset
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")

print(df.head())
print(df.info())
print(df.describe())

# Output Analysis
# no missing values, all data types are appropriate for analysis
# target variable is 'Final Grade' (numeric), and there are several potential predictor variables (numeric and categorical)
# Data Preprocessing
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)

   Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   
2              8               51                        Yes            7   
3              5               52                        Yes            5   
4              7               75                         No            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0  
<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                         

In [1]:
#Data Preprocessing
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

print(df.head())
print(df.info())

# creating feature matrix (X) and target vector (y)
X = df.drop("Performance Index", axis=1)
y = df["Performance Index"]

# verifying shapes of X and y
print(X.shape)
print(y.shape)

#Train-Test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# checking shapes of train and test sets
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

# linear regression model training
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

print("Weights:", model.coef_)
print("Bias:", model.intercept_)
for feature, weight in zip(X.columns, model.coef_):
    print(feature, ":", weight)

# model evaluation
from sklearn.metrics import mean_squared_error

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)

for i in range(5):
    print("Pred:", y_pred[i], "Actual:", y_test.values[i])

   Hours Studied  Previous Scores  Extracurricular Activities  Sleep Hours  \
0              7               99                           1            9   
1              4               82                           0            4   
2              8               51                           1            7   
3              5               52                           1            5   
4              7               75                           0            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0  
<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                   

In [2]:
# Feature Scaling
# since the features are on different scales, we will apply standardization to the feature matrix (X) before training the model
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

# creating feature matrix (X) and target vector (y)
X = df.drop("Performance Index", axis=1)
y = df["Performance Index"]

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

#Train-Test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# applying standardization to the feature matrix (X) before training the model
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# linear regression model training
from sklearn.linear_model import LinearRegression
model = LinearRegression()

model.fit(X_train_scaled, y_train)

for feature, weight in zip(X.columns, model.coef_):
    print(feature, ":", weight)

# model evaluation
y_pred = model.predict(X_test_scaled)

mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)

for i in range(5):
    print("Pred:", y_pred[i], "Actual:", y_test.values[i])

Hours Studied : 7.401340794046579
Previous Scores : 17.637271267177667
Extracurricular Activities : 0.304290755963927
Sleep Hours : 0.8100310701229938
Sample Question Papers Practiced : 0.5488417139901897
MSE: 4.082628398521843
Pred: 54.711853917490075 Actual: 51.0
Pred: 22.615512943389305 Actual: 20.0
Pred: 47.90314471125894 Actual: 46.0
Pred: 31.28976748030152 Actual: 28.0
Pred: 43.004570423805454 Actual: 41.0


### 📊 Student Performance Prediction — Multiple Linear Regression Analysis

**Objective:**
To understand how different factors affect student performance and build a regression model to predict the *Performance Index*.
---
### 🔹 Dataset Overview

The dataset contains the following features:

* Hours Studied
* Previous Scores
* Extracurricular Activities (categorical: Yes/No)
* Sleep Hours
* Sample Question Papers Practiced

Target variable:

* Performance Index

No missing values were found, and all features except *Extracurricular Activities* were numerical.

---

### 🔹 Data Preprocessing

* Converted *Extracurricular Activities* from categorical (Yes/No) to numerical (1/0).
* Separated features (X) and target (y).
* Performed an 80–20 train-test split to evaluate generalization.

---

### 🔹 Model Used

A **Multiple Linear Regression** model was trained to learn the relationship:

```
Performance Index = w₁·(Hours Studied) + w₂·(Previous Scores) + ... + b
```

---

### 🔹 Initial Observations (Without Scaling)

* Hours Studied appeared to have the highest weight.
* However, feature magnitudes were different (e.g., Previous Scores ranged up to 100), which can bias interpretation.

---

### 🔹 Feature Scaling

Applied **Standardization (StandardScaler)**:

* Transformed features to have mean = 0 and standard deviation = 1.
* This ensures all features contribute comparably.

---

### 🔹 Key Insights After Scaling

* **Previous Scores** emerged as the most influential feature.
  → Indicates that past academic performance is a strong predictor.

* **Hours Studied** is also highly important.
  → Confirms effort still plays a major role.

* **Sleep Hours** and **Practice Papers** have moderate influence.

* **Extracurricular Activities** had the least impact in this dataset.
  → Suggests limited correlation, though not necessarily lack of real-world importance.

---

### 🔹 Model Performance

* Mean Squared Error (MSE): ~4.08
* Root Mean Squared Error (RMSE): ~2

This means predictions are off by ~2 units on average, which is quite accurate given the target range.

---

### 🔹 Important Learnings

* Feature scaling does **not change model accuracy**, but significantly improves interpretability.
* Linear regression captures **correlations, not causation**.
* Proper feature representation is as important as the model itself.
* Real-world intuition must always be compared with model outputs.

---

### 🔹 Conclusion

The model successfully captures the relationship between study-related factors and performance.
The analysis highlights that **past performance and study effort** are the most significant predictors in this dataset.

Further improvements could include:

* Feature engineering (interaction terms, nonlinear features)
* Trying other models (logistic regression, decision trees)
* Evaluating overfitting and generalization further

---

*This experiment helped build a deeper understanding of feature importance, scaling, and model interpretation in practical ML workflows.*